csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -114,98 +114,561 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
-                                        out pinnedBufferHandles[inputIndex]);
+                input.ToNativeOnnxValue(out inputValuesArray[inputIndex], out pinnedInputBufferHandles[inputIndex], out disposeInputs[inputIndex]);
 
                 inputIndex++;
             }
 
-            string[] outputNamesArray = outputNames.ToArray();
-            IntPtr[] outputValueArray = new IntPtr[outputNames.Count];
+            // prepare outputs
+            string[] outputNamesArray = outputNames as string[] ?? outputNames.ToArray();
+            IntPtr[] outputValuesArray = new IntPtr[outputNames.Count];
 
             IntPtr status = NativeMethods.OrtRun(
-                                                this._nativeHandle,
+                                                _nativeHandle,
                                                 options.Handle,
-                                                inputNames,
-                                                inputTensors,
-                                                (UIntPtr)(inputTensors.Length),
+                                                inputNamesArray,
+                                                inputValuesArray,
+                                                (UIntPtr)inputs.Count,
                                                 outputNamesArray,
                                                 (UIntPtr)outputNames.Count,
-                                                outputValueArray /* An array of output value pointers. Array must be allocated by the caller */
+                                                outputValuesArray /* Empty array is passed in to receive output OrtValue pointers */
                                                 );
 
             try
             {
                 NativeApiStatus.VerifySuccess(status);
                 var result = new DisposableList<DisposableNamedOnnxValue>();
-                for (uint i = 0; i < outputValueArray.Length; i++)
+                for (uint i = 0; i < outputValuesArray.Length; i++)
                 {
-                    result.Add(DisposableNamedOnnxValue.CreateFromOnnxValue(outputNamesArray[i], outputValueArray[i]));
+                    result.Add(DisposableNamedOnnxValue.CreateFromOnnxValue(outputNamesArray[i], outputValuesArray[i]));
                 }
 
                 return result;
             }
             catch (OnnxRuntimeException e)
             {
                 //clean up the individual output tensors if it is not null;
-                for (uint i = 0; i < outputValueArray.Length; i++)
+                for (int i = 0; i < outputValuesArray.Length; i++)
                 {
-                    if (outputValueArray[i] != IntPtr.Zero)
+                    if (outputValuesArray[i] != IntPtr.Zero)
                     {
-                        NativeMethods.OrtReleaseValue(outputValueArray[i]);
+                        NativeMethods.OrtReleaseValue(outputValuesArray[i]);
```

Comment:
As an aside, can OrtReleasevalue() itself throw ? If it can/does, since there is no catch..., it seems like we will be leaking the other valid `DisposableNamedOnnxValue` s ? @jignparm ?

Reply:
All `OrtRelease***` C-APIs are implemented using C++ `delete`. Failure (if any) cannot be caught in managed code.